### ISSUE TITLE
Code size of generated packages is excessive.

### ISSUE THREAD

**Issue by zenhack** [zombiezen/go-capnproto2#34]
For the sandstorm schema, on linux amd64 `grain.a` is 19MiB (!). Many of the others are rather large as well. As a point of comparision, the largest package in the go standard library is `net/http`, the `.a` for which is ~3.7MiB.

One thing I have noticed is that the struct wrappers and accessor functions add up. I did some mesurments a while back and never got around to filing the bug. IIRC, each of these cost us one or two hundred KiB:

https://github.com/zenhack/go.sandstorm/blob/master/capnp/grain/grain.capnp.go#L15

Those got noticably smaller when I took out the if statement (it's pretty easy to imagine a compiler inlining it into a noop in that case, though I don't think it actually does). Not enough to solve the problem on it's own though.

I started to play with killing the struct wrappers and making types direct aliases of the field they're now wrapping, which IIRC would allow more functions to be effectively noops, but ran into issues with copying non-exported fields, which I didn't spend much time fussing with.

I'd like to visualize the sizes of various functions to get a sense of where the space is being taken up; probably more efficient than the futzing with schema and modifications to templates I was doing. Would need to research tooling for that; haven't dug into it much yet.

**Comment by zenhack** [comment]
To clarify, getting rid of te if statement and returning `PowerboxDescriptor{st}, err`.

**Comment by zombiezen** [comment]
Argh, I was wondering about this and hadn't gotten around to actually measuring it.  I can't imagine v1 is much better, but we do error checking, which probably messes with inlining.  Do you know if Go 1.6 versus Go 1.7 makes a difference?

**Comment by zenhack** [comment]
Haven't tried it with 1.7.

I the the specific case I mentioned, the error checking is actually redundant; we don't lose anything by just passing back whatever NewStruct gives us. I suspect there are a lot of cases like these.

**Comment by zombiezen** [comment]
Is this prohibitively bloating your binaries? I'm fairly sure that 1.7 will decrease the archive sizes because the export format is suboptimal.

Some poking around on my local machine.

```shell
$ go version
go version go1.6 linux/amd64
$ ls -lh ~/pkg/linux_amd64/zombiezen.com/go/capnproto2/std/capnp/
total 18M
-rw-r--r-- 1 light light  384 Jun  9 17:36 cxx.a
-rw-r--r-- 1 light light 989K Jun  9 17:39 json.a
-rw-r--r-- 1 light light 1.4M Jun  9 17:39 persistent.a
-rw-r--r-- 1 light light 5.6M Jun  9 17:39 rpc.a
-rw-r--r-- 1 light light 1.6M Jun  9 17:39 rpctwoparty.a
-rw-r--r-- 1 light light 7.7M Jun  9 17:39 schema.a
$ ls -lh `which capnpc-go`
-rwxr-xr-x 1 light light 8.1M Jun  9 17:39 /home/light/bin/capnpc-go
$ ls -lh hello  # a Hello World binary that pulls in fmt
-rwxr-xr-x 1 light light 2.2M Jun  9 17:44 hello
```

While I agree 8.1M is large, it's a 4x difference versus a 10x difference.  Not great, but okay.  I suspect that some of the SSA work in 1.7 makes this better.

**Comment by zombiezen** [comment]
If you're curious about debugging this, the two tools I've found handy are `go tool nm` and `go build -gcflags=-m`.  The former can spit out symbols and sizes in a library or executable, while the latter gives you insight into what can or can't be inlined.

There's not a smoking gun here, but I would have expected something like:

```go
func (s Method) SetImplicitParameters(v Node_Parameter_List) error {
  return s.Struct.SetPtr(4, v.List.ToPtr())
}
```

to be inlined, but the gc inliner will only inline calls to functions that themselves can be inlined.

**Comment by zombiezen** [comment]
I checked in 13585a71 which should reduce branching and size in most generated code, but not by a huge margin.

I ran the build against golang/go@8042bfe (tip).  Library sizes are roughly halved on disk, and the `capnpc-go` binary went from 8.1M to 6.4M.  Dead code elimination seems better.  Inlining behavior looks much more reasonable in Go 1.7: while the generated getters and setters are not marked inlinable, the calls they make to the underlying struct methods are being inlined.

**Comment by zenhack** [comment]
It hasn't actually become a practical issue yet.

Thanks for the tips re: tooling. nm was my first instinct, but got as far as "binutils nm doesn't understand go, I'll have to do some digging."

**Comment by zombiezen** [comment]
I'm going to close this as won't fix for now, since Go 1.7 will markedly improve the code size for no additional effort.  I'm making an effort to have new/optional features (like JSON and string marshalling) use the schema instead of code generation, which should keep the code size fairly constant over time.

If you find some ways to make the generated code more optimal, feel free to reopen.

**Comment by zenhack** [comment]
Fair enough.

Quoting Ross Light (2016-06-10 12:42:35)

>    I'm going to close this as won't fix for now, since Go 1.7 will
>    markedly improve the code size for no additional effort. I'm making an
>    effort to have new/optional features (like JSON and string marshalling)
>    use the schema instead of code generation, which should keep the code
>    size fairly constant over time.
>
>    If you find some ways to make the generated code more optimal, feel
>    free to reopen.
